### ISSUE TITLE
Discuss: App Configuration

### ISSUE THREAD

**Issue by djowel** [cycfi/elements#158]
Currently, Elements requires a `config.json` to be loaded on program startup. Right now, this file may seem to contain only frivolous information. But there's more to it than that. To be clear, the intent is that this config.json may contain a lot more startup information about the program, including, for example, UI strings, for things such as labels, titles, etc. The idea is that instead of hard-coding these strings, get it from the config file. That way, you can change the names, labels, etc. without having to rebuild the app, which makes it easy to localize and deploy. That's just one example. Another example is that it can store things such as paths to some more resources such as images, audio, etc., or how even fonts and theme information, etc...

On one hand, that is forward-looking, on the other hand, it may be too forward-looking and it can be argued that Elements should not care about such things and let the user decide on how his/her program manages such things.

So, I'm opening this "issue" for discussion. It is possible to remove the need for `config.json`. If so, now is a good time to do it, since it is early in the dev. stage and nothing depends on this "feature" yet.

Your thoughts?

**Comment by jpcima** [comment]
On the particular topic of localization:
there definitely exists a need for a localization system, for software like audio plugins.
I've discussed previously with some developers and a translator who wish there existed ready solutions for this particular use case.

The key->value model of string catalog is extremely basic, cuts it in some cases but sometimes not.
This is a system used by the JUCE framework. It's not able, among things, to deal with context-based resolution of ambiguity, plural forms, and it lacks a comfortable set of tools for maintaining string catalogs after code changes.

So basically there has been a floating idea to make a minimal GNU gettext alternative that could be fit to use in plugins. To my knowledge, this project has not yet been realized.

So in short, I'll be definitely in favor of a specialized system, not some kind of json key->value mapping.
And if this system exists, it had better be a dedicated separate library such that all developers are able to use it regardless of frameworks used.

**Comment by Xeverous** [comment]
> The key->value model of string catalog is extremely basic [...] It's not able, among things, to deal with context-based resolution of ambiguity, plural forms, and it lacks a comfortable set of tools for maintaining string catalogs after code changes.

I view the use of key->value system is as a simple but working solution to the "to uppercase/to lowercase" and many similar problems. Many languages have less or more than 2 casing states, some have more complex grammars, some have multiple plural/singular forms, some even more extra rules about diacritics or pre/su-fixes depending on the context. I don't see any way to fix that problem, human languages are just too different to make any mechanism that tries to make use of common parts (which are none across all languages). key-value mapping is a simple implementation.

If someone sees that in situation X language A requires extra context, it means more keys are needed and there is more than 1 X situation. Perhaps X and Y situations where language A requires 2 different strings but language B can use the same string.

> So in short, I'll be definitely in favor of a specialized system, not some kind of json key->value mapping. And if this system exists, it had better be a dedicated separate library such that all developers are able to use it regardless of frameworks used.

I agree with this. IMO elements tries too much on these things, this is a completely separate, non-trivial task and the GUI library should be just a library, with no framework-like imposed requirements. I would like to be able to use my own (separate) translation mechanism/library/whatever (if any).

> On one hand, that is forward-looking, on the other hand, it may be too forward-looking and it can be argued that Elements should not care about such things and let the user decide on how his/her program manages such things.

> So, I'm opening this "issue" for discussion. It is possible to remove the need for config.json. If so, now is a good time to do it, since it is early in the dev. stage and nothing depends on this "feature" yet.

I would remove it. I don't like any hardcoded paths or config structures and there is just too much variance to make any good implementation that would satisfy everyone. Just drop this feature from the library goals, as I wrote earlier: I prefer elements to be a GUI library, not GUI+config+translation+whatever-else-platform-specific-with-no-customization-that-satisfies-all. Code should be modular, and if someone wants a config/translation library, they should be free to use a separate library for their very unique requirements. IMO elements should not do any file operations besides stuff like "load font/image from this path" (where the path is explicitly given by the user).

**Comment by djowel** [comment]
OK, in light of all these, I'm in favor of removing config.json. 

Now here's a question: what's the reasonable way to inject runtime app specific information from cmake in Windows and Linux? The information needed are:

application_title
application_copyright
application_id
application_version

Probably more in the future. Not all these are currently used, but it would not hurt to have these. 

MacOS is good (it has the plist xml machinery). I need something for Linux and Windows. And, while we're at it, will you be interested in working on a PR? Perhaps @jpcima on Linux and @Xeverous on Windows?

Doing this step will remove one dependency.

P.S. Out of topic hint to @Xeverous ... if you want total independence from Boost, there's such a thing as ASIO standalone (header only!): https://think-async.com/Asio/AsioStandalone.html

**Comment by Xeverous** [comment]
> what's the reasonable way to inject runtime app specific information from cmake in Windows and Linux?

Why do you need this? I don't want to go that far with CMake (that is, coupling the source with CMake-provided information) - this complicates support for multiple build scripts. IMO this is going too far, it should be a runtime option specified in code.

> Probably more in the future. Not all these are currently used, but it would not hurt to have these.

I still don't get why do you need these. Windows implementation does not use any although it supplies some hardcoded window class name. Linux implementation uses only application ID which itself is optional since GTK 3.6.

Functions:

- https://developer.gnome.org/gtk3/stable/GtkApplication.html#gtk-application-new
- https://docs.microsoft.com/en-us/windows/win32/api/winuser/ns-winuser-wndclassw

Docs:

- https://developer.gnome.org/ChooseApplicationID/
- https://docs.microsoft.com/en-us/windows/win32/winmsg/about-window-classes

> I need something for Linux and Windows.

Honestly, Is it really needed? At most just 1 string is needed, that could be supplied through code.

**Comment by djowel** [comment]
> I still don't get why do you need these. Windows implementation does not use any although it supplies some hardcoded window class name. Linux implementation uses only application ID which itself is optional since GTK 3.6.

Ok, assuming I agree with you...

> Honestly, Is it really needed? At most just 1 string is needed, that could be supplied through code.

... then we still need at least one. How do you supply it "through code" in a cross-platform manner? Program options? Ugh. Hard-code? Double ugh! 

Also, keep in mind that MacOS does need it (info.plist).

**Comment by Xeverous** [comment]
> How do you supply it "through code" in a cross-platform manner?

`app` constructor? Isn't this the purpose of this class (cross-platform application/initialization abstraction)?

> Also, keep in mind that MacOS does need it (info.plist).

Does it need the file only? Or some data to pass at runtime too? If it's a file only, I think it is outside the scope of a GUI library - IMO the GUI should only be concerned in necessary information to just run the GUI.

**Comment by djowel** [comment]
> > How do you supply it "through code" in a cross-platform manner?
> 
> `app` constructor? Isn't this the purpose of this class (cross-platform application/initialization abstraction)?

OK, that's the easiest alright. I admit... I am over-thinking it :-)

**Comment by djowel** [comment]
> At most just 1 string is needed, that could be supplied through code.

Ah, Gtk also needs the application_id, which is something like "org.gtk.example". 

https://developer.gnome.org/gtk3/stable/GtkApplication.html#gtk-application-new

I think I'll just have the app-config struct as ctor param, with some reasonable defaults.

**Comment by djowel** [comment]
> Does it need the file only? Or some data to pass at runtime too? If it's a file only, I think it is outside the scope of a GUI library - IMO the GUI should only be concerned in necessary information to just run the GUI.

It needs the file only, and  it is included in the application bundle.

**Comment by djowel** [comment]
> application ID which itself is optional since GTK 3.6.

"If no application ID is given then some features (most notably application uniqueness) will be disabled." 

... better have it.

**Comment by Xeverous** [comment]
> It needs the file only, and it is included in the application bundle.

Right, so I think this is really outside the scope of the library. Linux has no such thing, at most distribution-specific package metadata and Windows ... well from nothing in old Win32 API to a whole application manifest for UWP and other "Windows app technologies".

**Comment by djowel** [comment]
> Right, so I think this is really outside the scope of the library.

Good point, actually. But I'll keep it in as it is as a benefit for users of the platform.

**Comment by petiaccja** [comment]
> ... then we still need at least one. How do you supply it "through code" in a cross-platform manner? Program options? Ugh. Hard-code? Double ugh!

I dislike hardcoding, too, but this is just one thing no matter how big the rest of the application is. Write once, never look at it again, so no maintenance burden even if not done nicely. The library only needs a string, so let's just take a string in the constructor, as @Xeverous said, and let the user determine how he will supply it -- your library is still as clean as it gets.

I also agree that metadata such as build numbers and copyrights do not belong to the GUI library. If someone has a different system, they will have to circumvent your library to do it their own way.

However, on Windows, I have used the window class name to detect if an instance of the application is already running. I think it would be useful if you included user-specified identifiers like this in the interface, but I can't speak for other platforms, so this is merely an observation.

**Comment by djowel** [comment]
Go check out the `no-boost-deps` branch. The branch is now free from all Boost dependencies:

https://github.com/cycfi/elements/tree/no-boost-deps

- JSON dependency is removed
- ASIO is now a standalone submodule in elements

There's an API change, obviously. I'll announce that, perhaps in a WIKI announcement page that links to the top readme.

**Comment by djowel** [comment]
This is now merged to develop and master. We're now free from all Boost dependencies. Closing this now.
